peps/pep-0767.rst

```diff
@@ -0,0 +1,553 @@
+type checkers perceive the mutability of the assigned object. Immutable :term:`ABCs <abstract base class>`
+and :mod:`containers <collections.abc>` may be used in combination with ``ReadOnly``
+to forbid mutation of such values:
+
+.. code-block:: python
+
+    from collections import abc
+    from dataclasses import dataclass
+    from typing import Protocol, ReadOnly
+
+
+    @dataclass
+    class Game:
+        name: str
+
+
+    class HasGames[T: abc.Collection[Game]](Protocol):
+        games: ReadOnly[T]
+
+
+    def add_games(shelf: HasGames[list[Game]]) -> None:
+        shelf.games.append(Game("Half-Life"))  # ok: list is mutable
+        shelf.games[-1].name = "Black Mesa"    # ok: "name" is not read-only
+        shelf.games = []                       # error: "games" is read-only
+        del shelf.games                        # error: "games" is read-only and cannot be deleted
+
+
+    def read_games(shelf: HasGames[abc.Sequence[Game]]) -> None:
+        shelf.games.append(...)             # error: "Sequence" has no attribute "append"
+        shelf.games[0].name = "Blue Shift"  # ok: "name" is not read-only
+        shelf.games = []                    # error: "games" is read-only
+
+
+All instance attributes of frozen dataclasses and ``NamedTuple`` should be
+implied to be read-only. Type checkers **should not** flag redundant annotations
+of such attributes with ``ReadOnly``:
+
+.. code-block:: python
+
+    from dataclasses import dataclass
+    from typing import NewType, ReadOnly
+
+
+    @dataclass(frozen=True)
+    class Point:
+        x: int            # implicit read-only
+        y: ReadOnly[int]  # ok, explicit read-only
+
+
+    uint = NewType("uint", int)
+
+
+    @dataclass(frozen=True)
+    class UnsignedPoint(Point):
+        x: ReadOnly[uint]  # ok, explicit & narrower type
+        y: uint            # ok, narrower type
+
+Initialization
+--------------
+
+Assignment to a read-only attribute can only occur in the class declaring the attribute.
+There is no restriction to how many times the attribute can be assigned to.
+The assignment must be allowed in the following contexts:
+
+1. In ``__init__``, on the instance received as the first parameter (likely, ``self``).
+2. In ``__new__``, on instances of the declaring class created via a call
+   to a super-class' ``__new__`` method.
+3. At declaration in the body of the class.
+
+Additionally, a type checker may choose to allow the assignment:
+
+1. In ``__new__``, on instances of the declaring class, without regard
+   to the origin of the instance.
+   (This choice trades soundness, as the instance may already be initialized,
+   for the simplicity of implementation.)
+2. In ``@classmethod``\ s, on instances of the declaring class created via
+   a call to the class' or super-class' ``__new__`` method.
+
+Note that a child class cannot assign to any read-only attribute of a parent class
+in any of the aforementioned contexts, unless the attribute is redeclared.
```

Comment:
@carljm how do you like this?

Reply:
I like it! We'll see if other reviewers agree :)

The code examples below are very clarifying for the other rules; I think it would be useful to expand them with examples of `__new__` and classmethods.

It may also be worth a section in "rejected alternatives" discussing _why_ being able to assign readonly attributes in classmethod constructors is useful (that is, the `Fraction` example.)

Reply:
What exactly would be the "rejected alternative"? Disallowing classmethods/assignment only in `__init__`?

Reply:
Yes, and the text would summarize the reasoning for allowing assignment in `__new__` and optionally other class methods.